### PR TITLE
Ticket646 lakeshore 460

### PR DESCRIPTION
**Pull request by esouthren**
### Description of work

IOC for the Lakeshore 460, part of the HIFI Cryomagnet software re-write

List of associated pull requests:
[Support files](https://github.com/ISISComputingGroup/EPICS-Lakeshore460/pull/1)
[OPI](https://github.com/ISISComputingGroup/ibex_gui/pull/559)
[Emulator](https://github.com/ISISComputingGroup/EPICS-DeviceEmulator/pull/42)
[IOC Testing framework tests](https://github.com/ISISComputingGroup/EPICS-IOC_Test_Framework/pull/45)
[IOC Master Makefile](https://github.com/ISISComputingGroup/EPICS-ioc/pull/209)
[EPICS](https://github.com/ISISComputingGroup/EPICS/pull/103)

Notes:

globals.txt settings
```
Hardware testing
LKSH460_01__DEVSIM=0
LKSH460_01__PORT=COM6

Emulator testing
LKSH460_01__DEVSIM=1
LKSH460_01__PORT=COM6
LKSH460_01__EMULATOR_PORT=57677

Unit Testing
#LKSH460_01__DEVSIM=1
LKSH460_01__PORT=COM6
#LKSH460_01__EMULATOR_PORT=57677
```

### Acceptance criteria

IOC works with Lakeshore 460 hardware; emulator emulates the device; OPI displays all required PVs.

---

#### Code Review

- [x] **Pertitent information and a copy of the manual has been stored in the [wiki](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Specific-Device-IOC)**
- [x] Does the IOC conform to IBEX standards?
    - [PV naming](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/PV-Naming).
    - [Disable records](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Disable-records)
    - [Record simulation](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Record-Simulation)
    - [Finishing touches](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/IOC-Finishing-Touches)
- [x] If an OPI has been modified, does it conform to the [style guidelines](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/OPI-Creation)?
- [x] Have the changes been recorded appropriately in the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev)?

### Functional Tests

- IOC responds correctly in:
    - [x] Devsim mode
    - [x] Recsim mode
    - [x] Real device, if available
- [x] Supplementary IOCs (`..._0n` where `n>1`) run correctly
- [x] Log files do not report undefined macros (serach for `macLib: macro` to find instances of `macLib: macro [macro name] is undefined...`

### Final steps

- [ ] Update the IOC submodule in the main EPICS repo. See [Git workflow](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Git-workflow) page for details.
- [x] Reviewer has moved the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev) entry for this ticket in the "Changes merged into master" section

### CODE REVIEW

File: LKSH460/iocBoot/iocLKSH460-IOC-01/st-common.cmd [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+< $(IOCSTARTUP)/init.cmd
+
+cd "${TOP}"
+
+# For dev sim devices
+$(IFDEVSIM) drvAsynIPPortConfigure("L0", "localhost:$(EMULATOR_PORT=57677)")
+
+## Register all support components
+dbLoadDatabase "dbd/LKSH460-IOC-01.dbd"
+LKSH460_IOC_01_registerRecordDeviceDriver pdbbase
```
Comment (Tom-Willemsen) [inline LKSH460/iocBoot/iocLKSH460-IOC-01/st-common.cmd]: This is duplicated in both `st.cmd` and here, which causes the following errors in the IOC log:

```
...
LKSH460_IOC_01_registerRecordDeviceDriver pdbbase
Warning: Registration already done.
registryFunctionRefAdd: could not register arrayTest_init
registryFunctionRefAdd: could not register asSubInit
registryFunctionRefAdd: could not register configMenuList_init
registryFunctionRefAdd: could not register interp_init
registryFunctionRefAdd: could not register initSubAve
Warning: iocshRegisterVariable redefining CASDEBUG.
Warning: iocshRegisterVariable redefining aCalcAsyncThreshold.
Warning: iocshRegisterVariable redefining aCalcLoopMax.
Warning: iocshRegisterVariable redefining aCalcPerformDebug.
Warning: iocshRegisterVariable redefining aCalcPostfixDebug.
Warning: iocshRegisterVariable redefining aCalcoutRecordDebug.
Warning: iocshRegisterVariable redefining arrayTestDebug.
Warning: iocshRegisterVariable redefining asCaDebug.
Warning: iocshRegisterVariable redefining atExitDebug.
Warning: iocshRegisterVariable redefining boHIGHlimit.
Warning: iocshRegisterVariable redefining boHIGHprecision.
Warning: iocshRegisterVariable redefining calcoutODLYlimit.
Warning: iocshRegisterVariable redefining calcoutODLYprecision.
Warning: iocshRegisterVariable redefining callbackParallelThreadsDefault.
Warning: iocshRegisterVariable redefining configMenuDebug.
Warning: iocshRegisterVariable redefining dbBptNotMonotonic.
Warning: iocshRegisterVariable redefining dbQuietMacroWarnings.
Warning: iocshRegisterVariable redefining dbRecordsAbcSorted.
Warning: iocshRegisterVariable redefining dbRecordsOnceOnly.
Warning: iocshRegisterVariable redefining dbTemplateMaxVars.
Warning: iocshRegisterVariable redefining dbThreadRealtimeLock.
Warning: iocshRegisterVariable redefining debugSubAve.
Warning: iocshRegisterVariable redefining devaCalcoutSoftDebug.
Warning: iocshRegisterVariable redefining devsCalcoutSoftDebug.
Warning: iocshRegisterVariable redefining histogramSDELprecision.
Warning: iocshRegisterVariable redefining interpDebug.
Warning: iocshRegisterVariable redefining sCalcLoopMax.
Warning: iocshRegisterVariable redefining sCalcPerformDebug.
Warning: iocshRegisterVariable redefining sCalcPostfixDebug.
Warning: iocshRegisterVariable redefining sCalcStackHW.
Warning: iocshRegisterVariable redefining sCalcStackLW.
Warning: iocshRegisterVariable redefining sCalcoutRecordDebug.
Warning: iocshRegisterVariable redefining save_restoreDatedBackupFiles.
Warning: iocshRegisterVariable redefining save_restoreDebug.
Warning: iocshRegisterVariable redefining save_restoreIncompleteSetsOk.
Warning: iocshRegisterVariable redefining save_restoreNumSeqFiles.
Warning: iocshRegisterVariable redefining save_restoreRemountThreshold.
Warning: iocshRegisterVariable redefining save_restoreSeqPeriodInSeconds.
Warning: iocshRegisterVariable redefining seqDLYlimit.
Warning: iocshRegisterVariable redefining seqDLYprecision.
Warning: iocshRegisterVariable redefining sseqRecDebug.
Warning: iocshRegisterVariable redefining streamDebug.
Warning: iocshRegisterVariable redefining swaitRecordDebug.
Warning: iocshRegisterVariable redefining transformRecordDebug.
...
```

File: LKSH460/LKSH460-IOC-01App/Db/lakeshore460.db [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	field(ONST, "XYZ")
+	field(TWST, "XY")
+	field(THST, "XZ")
+	field(FRST, "YZ")
+	field(FVST, "X-Y")
+	field(ZRVL, "1")
+	field(ONVL, "2")
+	field(TWVL, "3")
+	field(THVL, "4")
+	field(FRVL, "5")
```
[the remark this answers is not in the document]
Comment (esouthren) [reply LKSH460/LKSH460-IOC-01App/Db/lakeshore460.db]: Synced these up to match

File: LKSH460/LKSH460-IOC-01App/Db/lakeshore460.db [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+record(mbbi, "$(P)CHANNEL")
+{
+	field(DESC, "Get channel")
+	field(DTYP, "stream")
+	field(INP,  "@devLakeshore_460.proto getChannel $(PORT)")
+	field(ONST, "X")
+	field(TWST, "Y")
+	field(THST, "Z")
+	field(FRST, "V")
```
[the remark this answers is not in the document]
Comment (esouthren) [reply LKSH460/LKSH460-IOC-01App/Db/lakeshore460.db]: Added

File: LKSH460/LKSH460-IOC-01App/Db/lakeshore460.db [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+record(mbbo, "$(P)CHANNEL:SP")
+{
+	field(DESC, "Set Vector magnitude source")
```
[the remark this answers is not in the document]
Comment (esouthren) [reply LKSH460/LKSH460-IOC-01App/Db/lakeshore460.db]: Changed description

File: LKSH460/LKSH460-IOC-01App/Db/lakeshore460_channels.template [hunk 5]
```diff
@@ -0,0 +1,424 @@
...
+	field(SCAN, "1 second")
+	field(DTYP, "stream")
+	field(INP, "@devLakeshore_460.proto getRelModeSetpoint($(CHANNEL)) $(PORT)")
+	info(INTEREST, "HIGH")	
+	info(ARCHIVE, "VAL")
+	field(EGU, "")
+	field(PREC, "4")
+}
+
+record(ao, "$(P)$(CHANNEL):RELMODESETPOINT:SP")
```
[the remark this answers is not in the document]
Comment (esouthren) [reply LKSH460/LKSH460-IOC-01App/Db/lakeshore460_channels.template]: I'll rename it to RELMODESET. It's to set a PV which the manual calls "relative mode setpoint"

From the manual - "The relative function lets the user see small variations in larger fields. The setpoint (or center) of the relative reading is set with the Relative Set key."

File: LKSH460/LKSH460-IOC-01App/Db/lakeshore460_channels.template [hunk 6]
```diff
@@ -0,0 +1,424 @@
...
+
+record(stringout, "$(P)$(CHANNEL):FIELDREADING:UNIT:CALC:STR")
+{
+    field(DESC, "Push units to field reading")
+    field(DOL, "$(P)$(CHANNEL):FIELDREADING:UNIT:CALC.SVAL CP")
+	field(OMSL, "closed_loop")
+    field(OUT,  "$(P)$(CHANNEL):FIELDREADING.EGU PP")
+}
+
+### FLNK STARTS HERE ###
```
[the remark this answers is not in the document]
Comment (esouthren) [reply LKSH460/LKSH460-IOC-01App/Db/lakeshore460_channels.template]: Removed

File: LKSH460/iocBoot/iocLKSH460-IOC-01/st-common.cmd (hunk 1 above)
Comment (esouthren) [reply LKSH460/iocBoot/iocLKSH460-IOC-01/st-common.cmd]: Removed duplication

File: LKSH460/iocBoot/iocLKSH460-IOC-01/st-common.cmd [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+## For real device use:
+$(IFNOTDEVSIM) $(IFNOTRECSIM) drvAsynSerialPortConfigure("L0", "$(PORT=COM6)", 0, 0, 0, 0)
+$(IFNOTDEVSIM) $(IFNOTRECSIM) asynSetOption("L0", -1, "baud", "$(BAUD=9600)")
+$(IFNOTDEVSIM) $(IFNOTRECSIM) asynSetOption("L0", -1, "bits", 7)
```
[the remark this answers is not in the document]
Comment (esouthren) [reply LKSH460/iocBoot/iocLKSH460-IOC-01/st-common.cmd]: Done

File: LKSH460/iocBoot/iocLKSH460-IOC-01/st-common.cmd [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+$(IFNOTDEVSIM) $(IFNOTRECSIM) asynSetOption("L0", -1, "parity", "$(PARITY=odd)")
+$(IFNOTDEVSIM) $(IFNOTRECSIM) asynSetOption("L0", -1, "stop", "$(STOP=1)")
+
+asynSetTraceMask("L0",-1,0x9) 
+asynSetTraceIOMask("L0",-1,0x2)
```
[the remark this answers is not in the document]
Comment (esouthren) [reply LKSH460/iocBoot/iocLKSH460-IOC-01/st-common.cmd]: Done

File: LKSH460/iocBoot/iocLKSH460-IOC-02/st.cmd [hunk 9]
```diff
@@ -0,0 +1,38 @@
+#!../../bin/windows-x64/LKSH460-IOC-02
```
[the remark this answers is not in the document]
Comment (esouthren) [reply LKSH460/iocBoot/iocLKSH460-IOC-02/st.cmd]: Added correct call to st-common

[the remark this answers is not in the document]
Comment (esouthren) [reply LKSH460/iocBoot/iocLKSH460-IOC-02/st.cmd]: Need to add LKSH460_02 macros to globals.txt to run

File: Makefile [hunk 10]
```diff
@@ -12,7 +12,7 @@ IOCDIRS = AG33220A AG3631A AG53220A CCD100 CONEXAGP CONTROLSVCS CRYVALVE DELFTAR
 IOCDIRS += HAMEG8123 HIFIMAG HLG HVCAEN INHIBITR INSTETC INSTRON ISISDAE JULABO KHLY2400 KEPCO LINKAM95 LINMOT LKSH218 LKSH336 
 IOCDIRS += MCLEN MERCURY_ITC MK2CHOPR MK3CHOPR NANODAC NEOCERA PDR2000 PIMOT PSCTRL 
 IOCDIRS += RUNCTRL SCIMAG3D SDTEST SKFCHOPPER SMC100 SPINFLIPPER306015 STPS350 STSR400 TDK_LAMBDA_GENESYS TEKAFG3XXX TEKDMM40X0 TEKMSO4104B TEST TPG26x TPG300 TTIEX355P BKHOFF
-IOCDIRS += ROTSC AMINT2L SPRLG FERMCHOP SAMPOS RKNPS CYBAMAN EGXCOLIM IEG
+IOCDIRS += ROTSC AMINT2L SPRLG FERMCHOP SAMPOS RKNPS CYBAMAN EGXCOLIM IEG LKSH460 KHLY2700
```
[the remark this answers is not in the document]
Comment (esouthren) [reply Makefile]: removed

File: LKSH460/LKSH460-IOC-01App/Db/lakeshore460_channels.template [hunk 11]
```diff
@@ -0,0 +1,424 @@
...
+record(ai, "$(P)$(CHANNEL):FWIN")  
+{
+    field(DESC, "Display filter window")
+    field(DTYP, "stream")
+    field(INP,  "@devLakeshore_460.proto getFWin($(CHANNEL)) $(PORT)")
+	field(HIHI, "10")
+	field(LOLO, "1")
+	field(HHSV, "MAJOR")
+    field(LLSV, "MAJOR")
+	field(FLNK, "$(P)$(CHANNEL):FNUM")
```
[the remark this answers is not in the document]
Comment (esouthren) [reply LKSH460/LKSH460-IOC-01App/Db/lakeshore460_channels.template]: Added

**Comment by esouthren** [comment]
Submitted for rework

File: HLG/iocBoot/iocHLG-IOC-01/st-common.cmd [hunk 12]
```diff
@@ -7,7 +7,7 @@ epicsEnvSet "STREAM_PROTOCOL_PATH" "$(HLG)/data"
 $(IFRECSIM) drvAsynSerialPortConfigure("L0", "$(PORT=NUL)", 0, 1, 0, 0)
 
 # For dev sim devices
-$(IFDEVSIM) drvAsynIPPortConfigure("L0", "localhost:$(EMULATOR_PORT=)")
+$(IFDEVSIM) drvAsynIPPortConfigure("L0", "localhost:$(EMULATOR_PORT=57677)")
```
Comment (davidkeymer) [inline HLG/iocBoot/iocHLG-IOC-01/st-common.cmd]: Did you really mean to change the HLG?

Comment (esouthren) [reply HLG/iocBoot/iocHLG-IOC-01/st-common.cmd]: Not sure how that IOC got put in my commit :/

Comment (esouthren) [reply HLG/iocBoot/iocHLG-IOC-01/st-common.cmd]: Removed that file.

File: LKSH460/iocBoot/iocLKSH460-IOC-01/config.xml [hunk 13]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" ?>
+<ioc_config xmlns:xi="http://www.w3.org/2001/XInclude">
+<config_part>
+<ioc_desc>Lakeshore460 </ioc_desc>
+<macros>
+<macro name="PORT" pattern="^COM[0-9]+$" description="Serial COM Port" />
+<macro name="BAUD" pattern="^[0-9]+$" description="Serial communication baud rate, defaults to 9600." />
+<macro name="BITS" pattern="^[0-9]$" description="Serial communication number of bits, defaults to 7." />
+<macro name="PARITY" pattern="^(even)|(odd)|(none)$" description="Serial communication parity, defaults to even." />
```
[the remark this answers is not in the document]
Comment (esouthren) [reply LKSH460/iocBoot/iocLKSH460-IOC-01/config.xml]: Changed

**Pull request merged**